src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,455 @@
+	       "lib_manager_get_instance_bss_address() instance_bss_size: 0x%x, pointer: 0x%x",
+			instance_bss_size, va_base);
+	return va_base;
+}
+
+int lib_manager_allocate_module_instance(uint32_t module_id, uint32_t instance_id,
+					 uint32_t is_pages, struct sof_man_module *mod)
+{
+	uint32_t bss_size =
+			(mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count)
+			 * HOST_PAGE_SIZE;
+	uint8_t *va_base = lib_manager_get_instance_bss_address(module_id, instance_id, mod);
+
+	if ((is_pages * HOST_PAGE_SIZE) > bss_size) {
+		tr_err(&lib_manager_tr,  "is_pages (%d) invalid, required: %d",
+		       is_pages, bss_size/HOST_PAGE_SIZE);
+		return -ENOMEM;
+	}
+	/*
+	 * TODO: Map bss memory here when mm_map() available.
+	 * And clear BSS
+	 */
+	va_base = va_base;
+	return 0;
+}
+
+int lib_manager_allocate_module(const struct comp_driver *drv,
+				struct comp_ipc_config *ipc_config,
+				void *ipc_specific_config,
+				uint32_t *module_entry_point)
+{
+	struct sof_man_fw_desc *desc;
+	struct sof_man_module *mod;
+	struct ipc4_base_module_cfg *base_cfg = ipc_specific_config;
+	int ret = -ENOMEM;
+	uint32_t module_id = IPC4_MOD_ID(ipc_config->id);
+	uint32_t entry_index = LIB_MANAGER_GET_MODULE_INDEX(module_id);
+
+	/* TODO: Now module is in HP SRAM. If IMR will be used, here we need to copy it */
+	/*       from external MEMORY to HP SRAM. */
+	desc = lib_manager_get_library_module_desc(module_id);
+	/* TODO: here we should validate module manifest, map memory if it is in external MEMORY */
+	mod = (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(entry_index));
+
+	ret = lib_manager_allocate_module_instance(module_id, IPC4_INST_ID(ipc_config->id),
+			base_cfg->is_pages, mod);
+	if (ret < 0) {
+		tr_err(&lib_manager_tr, "lib_manager_allocate_module() failed: %d", ret);
+		return ret;
+	}
+	*module_entry_point = mod->entry_point;
+	return ret;
+}
+
+void lib_manager_init(void)
+{
+	struct sof *my_sof = sof_get();
+
+	my_sof->ext_library = &loader_ext_lib;
+}
+
+struct sof_man_fw_desc *lib_manager_get_library_module_desc(int module_id)
+{
+	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	struct ext_library *_ext_lib = ext_lib_get();
+	char *buffptr = (char *)_ext_lib->desc[lib_id];
+
+	return (struct sof_man_fw_desc *)(buffptr + SOF_MAN_ELF_TEXT_OFFSET);
+}
+
+void lib_manager_update_sof_ctx(struct sof_man_fw_desc *desc, uint32_t lib_id)
+{
+	struct ext_library *_ext_lib = ext_lib_get();
+
+	_ext_lib->desc[lib_id] = desc;
+	/* TODO: maybe need to call here dcache_writeback here? */
+}
+
+int lib_manager_register_module(struct sof_man_fw_desc *desc, int module_id)
+{
+	/* allocate new  comp_driver_info */
+	struct comp_driver_info *new_drv_info;
+	struct comp_driver *drv;
+	struct sof_man_module *mod;
+	int ret = -ENOMEM;
+
+	uint32_t align = 4;
+	uint32_t entry_index = LIB_MANAGER_GET_MODULE_INDEX(module_id);
+
+	new_drv_info = (struct comp_driver_info *)rballoc_align(0, SOF_MEM_CAPS_RAM,
```

Comment:
no need for a type-cast: allocations return `void *`

Reply:
sorry, what kind of compilation crash? This really shouldn't be needed. Check all other calls to this function in the code. Particularly makes no sense at all to type-cast it to `(void *)`. Where you type-cast it to an integer - that I can understand, that the compiler would complain about.

Reply:
To be sure, do you mean to remove (void *) casts not (struct comp_driver_info *) as it is showed in github?

Reply:
all of them